pkg/query-service/app/clickhouseReader/reader.go

```diff
@@ -2753,3 +2762,289 @@ func (r *ClickHouseReader) GetSamplesInfoInLastHeartBeatInterval(ctx context.Con
+		err := r.db.Exec(ctx, query)
+		if err != nil {
+			return &model.ApiError{Err: err, Typ: model.ErrorInternal}
+		}
+	} else {
+		// remove index
+		query := fmt.Sprintf("ALTER TABLE %s.%s DROP INDEX IF EXISTS %s_idx", r.logsDB, r.logsTable, field.Name)
+		err := r.db.Exec(ctx, query)
+		if err != nil {
+			return &model.ApiError{Err: err, Typ: model.ErrorInternal}
+		}
+	}
+	return nil
+}
+
+func (r *ClickHouseReader) GetLogs(ctx context.Context, params *model.LogsFilterParams) (*[]model.GetLogsResponse, *model.ApiError) {
+	response := &[]model.GetLogsResponse{}
+	fields, apiErr := r.GetLogFields(ctx)
+	if apiErr != nil {
+		return nil, apiErr
+	}
+
+	filterSql, err := logs.GenerateSQLWhere(fields, params)
+	if err != nil {
+		return nil, &model.ApiError{Err: err, Typ: model.ErrorBadData}
+	}
+
+	query := fmt.Sprintf("SELECT "+
+		"timestamp, id, trace_id, span_id, trace_flags, severity_text, severity_number, body,"+
+		"CAST((attributes_string_key, attributes_string_value), 'Map(String, String)') as  attributes_string,"+
+		"CAST((attributes_int64_key, attributes_int64_value), 'Map(String, Int64)') as  attributes_int64,"+
+		"CAST((attributes_float64_key, attributes_float64_value), 'Map(String, Float64)') as  attributes_float64,"+
+		"CAST((resources_string_key, resources_string_value), 'Map(String, String)') as resources_string "+
+		"from %s.%s", r.logsDB, r.logsTable)
+
+	if filterSql != nil && *filterSql != "" {
+		query += fmt.Sprintf(" where %s", *filterSql)
+	}
+
+	query = fmt.Sprintf("%s order by %s %s limit %d", query, params.OrderBy, params.Order, params.Limit)
+	zap.S().Debug(query)
+	err = r.db.Select(ctx, response, query)
+	if err != nil {
+		return nil, &model.ApiError{Err: err, Typ: model.ErrorInternal}
+	}
+
+	return response, nil
+}
+
+func (r *ClickHouseReader) TailLogs(ctx context.Context, client *model.LogsTailClient) {
+
+	fields, apiErr := r.GetLogFields(ctx)
+	if apiErr != nil {
+		client.Error <- apiErr.Err
+		return
+	}
+
+	filterSql, err := logs.GenerateSQLWhere(fields, &model.LogsFilterParams{
+		Query: client.Filter.Query,
+	})
+
+	if err != nil {
+		client.Error <- err
+		return
+	}
+
+	query := fmt.Sprintf("SELECT "+
+		"timestamp, id, trace_id, span_id, trace_flags, severity_text, severity_number, body,"+
+		"CAST((attributes_string_key, attributes_string_value), 'Map(String, String)') as  attributes_string,"+
+		"CAST((attributes_int64_key, attributes_int64_value), 'Map(String, Int64)') as  attributes_int64,"+
+		"CAST((attributes_float64_key, attributes_float64_value), 'Map(String, Float64)') as  attributes_float64,"+
+		"CAST((resources_string_key, resources_string_value), 'Map(String, String)') as resources_string "+
+		"from %s.%s", r.logsDB, r.logsTable)
+
+	currentTime := uint64(time.Now().UnixNano() / int64(time.Millisecond))
```

Comment:
You wouldn't have to do this `uint64` conversion if the DB were to have signed timestamps?

Reply:
Yes correct 😅